### PR TITLE
Updating HttpMethodMetadata based on ICorsMetadata

### LINKED ISSUE #42994: Cors not working in Minimal API endpoints
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

I have a client application that send a `DELETE` request to a Minimal API endpoint hosted in an application. Since it is a Cross-Origin request, the browser tries to send a preflight request `Options` that fails even when my application is configured correctly to allow `Origin`, `Method` and `Headers`.

`Access to fetch at 'https://localhost:7083/api/category/7' from origin 'https://localhost:7280' has been blocked by CORS policy: Response to preflight request doesn't pass access control check: No 'Access-Control-Allow-Origin' header is present on the requested resource. If an opaque response serves your needs, set the request's mode to 'no-cors' to fetch the resource with CORS disabled. `

Based on the logs, listed below, the `CORS` policy is not found and a `405 HTTP Method Not Supported` response is generated. 

```
info: Microsoft.AspNetCore.Hosting.Diagnostics[1]
      Request starting HTTP/2 OPTIONS https://localhost:7083/api/category/7 - -
info: Microsoft.AspNetCore.Cors.Infrastructure.CorsMiddleware[10]
      No CORS policy found for the specified request.
info: Microsoft.AspNetCore.Routing.EndpointMiddleware[0]
      Executing endpoint '405 HTTP Method Not Supported'
info: Microsoft.AspNetCore.Routing.EndpointMiddleware[1]
      Executed endpoint '405 HTTP Method Not Supported'
info: Microsoft.AspNetCore.Hosting.Diagnostics[2]
      Request finished HTTP/2 OPTIONS https://localhost:7083/api/category/7 - - - 405 0 - 2.8323ms
```

I did some investigation and the `fake` `405 HTTP Method Not Supported` is generated by the `EndpointRoutingMiddleware`, not CorsMiddleware, specifically the `HttpMethodMatcherPolicy`. 

It is happening because when the `MapXXX` methods are called, a `HttpMethodMetadata` is added to the endpoint, however the `AcceptCorsPreflight` is not set (default: `false`).

`MVC/API Controllers` suffers the same problem, however the call to `AddCors` adds an `IApplicationModelProvider` that changes all `HttpMethodMetadata's` `AcceptCorsPreflight ` properties to `true`.

https://github.com/dotnet/aspnetcore/blob/6df6573d746f4a3b12cea51af9a15bfab9e0743a/src/Mvc/Mvc.Cors/src/CorsApplicationModelProvider.cs#L147

### Expected Behavior

Since my application is configured correctly, the `OPTIONS` request should completed successfully and set the required headers and the browser will send the `DELETE` request.

### Steps To Reproduce

``` c#
const string CorsPolicyName = "_myCorsPolicy";

var builder = WebApplication.CreateBuilder(args);
builder.Services.AddCors(options 
  => options.AddPolicy(name: CorsPolicyName, builder => builder.AllowAnyOrigin().AllowAnyHeader().AllowAnyMethod()));

var app = builder.Build();

app.UseHttpsRedirection();
app.UseCors();

var apis = app.MapGroup("/api").RequireCors(CorsPolicyName);
apis.MapDelete("category/{id}", async (int id) => Results.Ok());
```

### Exceptions (if any)

_No response_

### .NET Version

7.0.100-preview.7.22370.3

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by brunolins16**
Fixes #42994

### CODE REVIEW

File: src/Http/Routing/src/RouteEndpointBuilder.cs
Comment (BrennanConroy) [inline src/Http/Routing/src/RouteEndpointBuilder.cs:111]: Should DisableCors set `AcceptCorsPreflight` to true? Does the cors middleware have behavior we'd prefer over the fake 405 endpoint in that case?

Comment (brunolins16) [reply src/Http/Routing/src/RouteEndpointBuilder.cs:111]: > Should DisableCors set AcceptCorsPreflight to true?

I was concerned about that, however that is basically how MVC detects the CORS metadata.

> Does the cors middleware have behavior we'd prefer over the fake 405 endpoint in that case?

Not sure if preferable or not (MVC endpoints will no produce 405 in this case), but yes the middleware handle the `IDisableCorsAttribute` and complete the `OPTIONS` request with `204 NO Content` without the expected headers

Comment (brunolins16) [reply src/Http/Routing/src/RouteEndpointBuilder.cs:111]: Also, since we don't have `CORS` attributes in the routing layer, we could not determine if it is `Disable` or `Enable` at this layer :(.

Comment (brunolins16) [reply src/Http/Routing/src/RouteEndpointBuilder.cs:111]: Also, in my opinion make more sense the routing delegates the decision to fail the `OPTIONS` request to the CORS middleware, since CORS is present. And only fail as 405 when it is truly not able to handle (Not using CORS). Is that make sense?

File: src/Http/Routing/src/HttpMethodMetadata.cs
Comment (davidfowl) [inline src/Http/Routing/src/HttpMethodMetadata.cs:50]: internal? 😢

File: src/Http/Routing/src/RouteEndpointBuilder.cs
Comment (davidfowl) [inline src/Http/Routing/src/RouteEndpointBuilder.cs:91]: Is this the right place to do this?

File: src/Http/Routing/src/HttpMethodMetadata.cs
Comment (halter73) [reply src/Http/Routing/src/HttpMethodMetadata.cs:50]: I'm sad about this too. The other option @brunolins16 showed me first was to look for the last `IHttpMethodMetadata` no matter what the type. And then add a new `HttpMethodMetadata` with the methods copied and `AcceptCorsPreflight` set to true. I thought this was worse though, because now the custom `IHttpMethodMetadata` type is no longer *the* `IHttpMethodMetadata`.

Can we add a default setter implementation to the interface that throws? That at least in theory would be extensible.

File: src/Http/Routing/src/RouteEndpointBuilder.cs
Comment (halter73) [reply src/Http/Routing/src/RouteEndpointBuilder.cs:91]: I was debating this too. The other option is RouteEndpointDataSource. We need to make sure that this isn't redoing what MVC is doing already too. Maybe we can consolidate the MVC logic here though. And it mightfix other EndpointDataSources for free. I'm leaning towards leaving it hear and removing the special logic in MVC.

Comment (brunolins16) [reply src/Http/Routing/src/RouteEndpointBuilder.cs:91]: I don't know much about routing yet but seems a good idea to add it here and have other datasources have it fixed for free, Is there any other place where we could have it as common behavior?

Comment (brunolins16) [reply src/Http/Routing/src/RouteEndpointBuilder.cs:91]: Interesting, let me do some tests with MVC to see how it behave. The `CorsApplicationModelProvider` could not be remove because it does more than that but for sure we could remove this part.

File: src/Http/Routing/src/HttpMethodMetadata.cs
Comment (brunolins16) [reply src/Http/Routing/src/HttpMethodMetadata.cs:50]: So, will it require an API change, right? Let me work on it to have it ready for next API review.

Comment (brunolins16) [reply src/Http/Routing/src/HttpMethodMetadata.cs:50]: Is the suggestion to look for the last `IHttpMethodMetadata`, instead of the `HttpMethodMetadata`? If so, that means a custom `IHttpMethodMetadata` by default will throw when we set it in the new method, is that what you are saying?

File: src/Http/Routing/src/RouteEndpointBuilder.cs
Comment (brunolins16) [reply src/Http/Routing/src/RouteEndpointBuilder.cs:91]: > Interesting, let me do some tests with MVC to see how it behave. The `CorsApplicationModelProvider` could not be remove because it does more than that but for sure we could remove this part.

I like it because the `CorsApplicationModelProvider` implementation does not work for something like this:

``` c#
// Configure the HTTP request pipeline.
app.MapControllers().RequireCors("t");
```

File: src/Http/Routing/src/HttpMethodMetadata.cs
Comment (brunolins16) [reply src/Http/Routing/src/HttpMethodMetadata.cs]: Changing to add a default implementation (we will need add to both setter and getter :( ) make the property hidden for custom implementations, no hint at all in the intellisense. To be honest, I am not sure if that is good.☹️

Comment (brunolins16) [reply src/Http/Routing/src/HttpMethodMetadata.cs]: > The other option @brunolins16 showed me first was to look for the last IHttpMethodMetadata no matter what the type. And then add a new HttpMethodMetadata with the methods copied and AcceptCorsPreflight set to true. I thought this was worse though, because now the custom IHttpMethodMetadata type is no longer the IHttpMethodMetadata.

I know @halter73 does not like it but I still prefer it.  I know that it could potentially break someone using `endpoint.Metadata.GetMetadata<IHttpMethodMetadata> as MyCustomMetadata`, however, expecting that the most significant metadata is the custom type is risky already because there is no guarantee that any other code added a metadata after it.

Comment (halter73) [reply src/Http/Routing/src/HttpMethodMetadata.cs]: > Is the suggestion to look for the last `IHttpMethodMetadata`, instead of the `HttpMethodMetadata`? If so, that means a custom `IHttpMethodMetadata` by default will throw when we set it in the new method, is that what you are saying?

Yes. I think the intersection of CORS and custom IHttpMethodMetadata is small, and we can make a clear exception message that says that the setter needs to be implemented.

> Changing to add a default implementation (we will need add to both setter and getter :( ) make the property hidden for custom implementations, no hint at all in the intellisense. To be honest, I am not sure if that is good.☹️

IHttpMethodMetadata.AcceptCorsPreflight is not an API that many ASP.NET Core developers are looking at with intellisense. It's an advanced API. As long as the exception is clear, I think we're fine.

File: src/Http/Routing/src/RouteEndpointBuilder.cs
Comment (halter73) [inline src/Http/Routing/src/RouteEndpointBuilder.cs:119]: Should we also try to avoid potential issues if `AcceptCorsPreflight` is already true? If we do this, we should also add a regression test for the behavior.

```suggestion
            if (hasCorsMetadata && httpMethodMetadata is not null && !httpMethodMetadata.AcceptCorsPreflight)
```

File: src/Mvc/Mvc.Cors/test/CorsApplicationModelProviderTest.cs
Comment (halter73) [inline src/Mvc/Mvc.Cors/test/CorsApplicationModelProviderTest.cs:23]: I assume we have some end-to-end tests covering this scenario. Do you know where they are?

**Review by halter73 — APPROVED** [review]

File: src/Http/Routing/src/RouteEndpointBuilder.cs
Comment (JamesNK) [inline src/Http/Routing/src/RouteEndpointBuilder.cs:123]: Metadata is supposed to be immutable. Rather than modifying the HTTP method metadata, have you considered checking for ICorsMetadata inside the HTTP method policy? If it's present, then go ahead with accepting preflight.

Comment (Kahbazi) [reply src/Http/Routing/src/RouteEndpointBuilder.cs:123]: Unrelated to this PR but while we are on subject of immutable metadata. `IAuthorizeData` has public setters for its properties. Wondering if it should be deprecated and removed though it would be a breaking change.

Comment (JamesNK) [reply src/Http/Routing/src/RouteEndpointBuilder.cs:123]: Metadata doesn't have to be immutable, and it's not worth breaking changes. It's a pattern to try to encourage.

Comment (brunolins16) [reply src/Http/Routing/src/RouteEndpointBuilder.cs:123]: > Metadata is supposed to be immutable. Rather than modifying the HTTP method metadata, have you considered checking for ICorsMetadata inside the HTTP method policy? If it's present, then go ahead with accepting preflight.

@JamesNK that was my first idea but after talking with @halter73 the suggestion to mutate the metadata sounded reasonable

Comment (brunolins16) [reply src/Http/Routing/src/RouteEndpointBuilder.cs:123]: @halter73 I can easily make the the changes in Http method policy without need a public api, do you have any concerns?

Comment (brunolins16) [reply src/Http/Routing/src/RouteEndpointBuilder.cs:123]: The only behavior that I don't like about not mutating the metadata, change only the policy, is that the `AcceptsCorsPreFlight` will never be true, so, there is no reason for it to exist.

Comment (halter73) [reply src/Http/Routing/src/RouteEndpointBuilder.cs:123]: @JamesNK Interesting idea. I also don't love mutating the metadata. We had a conversation about how to avoid we tried to this earlier in https://github.com/dotnet/aspnetcore/pull/43077#discussion_r937276637 by replacing the metadata, but that has other issues of course.

Having the CORS middleware handle this itself is another approach that I like. My problem with it is that:

> The only behavior that I don't like about not mutating the metadata, change only the policy, is that the AcceptsCorsPreFlight will never be true, so, there is no reason for it to exist.

If we go this route, we need to obsolete `AcceptCorsPreflight` and tell people to add the appropriate attribute the the endpoints metadata instead. I don't have much of a preference. As you say, we do have other mutable metadata, so we're not breaking entirely new ground here.

I've already approved this PR and I still think it's okay as is, but if people want to obsolete `AcceptCorsPreflight` instead, I'm okay with that too. That will need to be a new API proposal, but we could probably get that approved over email this afternoon if we think it's worth it.

File: src/Mvc/Mvc.Cors/test/CorsApplicationModelProviderTest.cs
Comment (brunolins16) [reply src/Mvc/Mvc.Cors/test/CorsApplicationModelProviderTest.cs:23]: Here:
https://github.com/dotnet/aspnetcore/blob/9af007591a271f176cc74f374d2c239908f1ac7d/src/Mvc/test/Mvc.FunctionalTests/CorsTestsBase.cs#L11

https://github.com/dotnet/aspnetcore/blob/9af007591a271f176cc74f374d2c239908f1ac7d/src/Mvc/test/Mvc.FunctionalTests/CorsEndpointRoutingTests.cs#L6

File: src/Http/Routing/src/RouteEndpointBuilder.cs
Comment (JamesNK) [reply src/Http/Routing/src/RouteEndpointBuilder.cs:123]: I don't have enough time before code complete to think and test this through to offer a strong opinion 😢

Comment (brunolins16) [reply src/Http/Routing/src/RouteEndpointBuilder.cs:96]: Ok, if not objections I am planning to merge the PR as is. It is broken since 3.x, so I feel like it is a must to have change.

**Comment by prkemp** [comment]
Can this change be merged into .net 6 for a subsequent release?  I'm interested in the fix for issue "Cors not working in Minimal API endpoints" #42994.   I'm not keen to upgrade to .net 7 yet.  Thanks

**Comment by BrennanConroy** [comment]
There is a workaround for .net 6, add `.WithMetadata(new HttpMethodMetadata(new string[] { "POST" }, acceptCorsPreflight: true));` to any affected endpoints.

**Comment by prkemp** [comment]
Great, thanks for this.

**Pull request merged**
